atlas/transactions/admin/fulfill_pack_buyback_offer.cdc

```diff
@@ -0,0 +1,114 @@
+        // Initialize NFT IDs and buyback prices
+        self.nftIDs = [{{.nftIDs}}]
+        self.prices = [{{.prices}}]
+
+        // Get or create user's storefront
+        if user.storage.borrow<&NFTStorefront.Storefront>(from: NFTStorefront.StorefrontStoragePath) == nil {
+            user.storage.save(<- NFTStorefront.createStorefront(), to: NFTStorefront.StorefrontStoragePath)
+            user.capabilities.publish(
+                user.capabilities.storage.issue<&NFTStorefront.Storefront>(NFTStorefront.StorefrontStoragePath),
+                at: NFTStorefront.StorefrontPublicPath
+            )
+        }
+        self.userStorefront = user.storage.borrow<auth(NFTStorefront.CreateListing) &NFTStorefront.Storefront>(from: NFTStorefront.StorefrontStoragePath)
+            ?? panic("Missing NFTStorefront")
+
+        // Get or create user's NFT withdrawal capability
+        let providerPath = /storage/{{.NFTProductName}}NFTCollectionProviderForNFTStorefront
+        self.userNFTWithdrawCap = user.storage.copy<Capability<auth(NonFungibleToken.Withdraw) &{{.NFTProductName}}.Collection>>(from: providerPath)
+        if self.userNFTWithdrawCap == nil {
+            self.userNFTWithdrawCap = user.capabilities.storage.issue<auth(NonFungibleToken.Withdraw) &{{.NFTProductName}}.Collection>({{.NFTProductName}}.CollectionStoragePath)
+            user.capabilities.storage.getController(byCapabilityID: self.userNFTWithdrawCap!.id)!.setTag("{{.NFTProductName}}CollectionProviderForNFTStorefront")
+            user.storage.save(self.userNFTWithdrawCap!, to: providerPath)
+        }
+
+        // Get user's DUC receiver capability
+        self.userDUCReceiverCap = user.capabilities.get<&{FungibleToken.Receiver}>(/public/dapperUtilityCoinReceiver)!
+        assert(self.userDUCReceiverCap.borrow() != nil, message: "Missing DUC receiver cap")
+
+        // Borrow Dapper's DUC vault
+        self.dapperVault = dapper.storage.borrow<auth(FungibleToken.Withdraw) &DapperUtilityCoin.Vault>(from: /storage/dapperUtilityCoinVault)
+            ?? panic("Missing Dapper DUC vault")
+
+        // Record Dapper's DUC balance
+        self.initialDapperBalance = self.dapperVault.balance
+
+        // Record issuer's address
+        self.issuerAddress = issuer.address
+
+        // Borrow issuer's NFT collection
+        self.issuerCollection = getAccount(self.issuerAddress).capabilities.borrow<&NonFungibleToken.Collection>(from: {{.NFTProductName}}.CollectionPublicPath)
+            ?? panic("Missing issuer NFT collection")
+    }
+
+    pre {
+        self.nftIDs.length == self.prices.length: "NFTs/prices length mismatch"
+        self.issuerAddress == 0x{{.NFTContractAddress}}: "Must be NFT contract owner"
+    }
+
+    execute {
+        // Gather active listings in user's storefront
+        let listingIDsByNFTID: {UInt64: UInt64} = {}
+        for id in self.userStorefront.getListingIDs() {
```

Comment:
this is to check if the user already listed the nft?

Reply:
i'm concerned about the for loop and `getListingIDs` for accounts with lots of listings, it could cause complexity limit errors. What if we discard their existing listings and just relist it and buy our own listing?

Reply:
Would we need to show users if a listing exists before they complete a buyback offer tho? We'd need to use platform-services until atlas is backfilled, probably not a blocker tho